### PR TITLE
Get rid of FunctionDefinition::UID and reuse function evaluation code

### PR DESCRIPTION
**Pull request by emmanuel099**
* Use the memory address of function definitions as UIDs instead of inventing another one
* Use existing function evaluation code to get the update location
    * Reduces code duplication
    * Makes it possible to update "method functions" (better name required ..) when the other bits are implemented

### CODE REVIEW

File: src/execute/NumericExecutionPass.cpp [hunk 1]
```diff
@@ -1238,45 +1257,15 @@ void ExecutionVisitor::visit( UpdateRule& node )
...
-        argumentValues.emplace_back( value );
-    }
-
-    assert( node.function()->targetType() == DirectCallExpression::TargetType::FUNCTION );
-    const auto& functionDefintion =
-        std::static_pointer_cast< FunctionDefinition >( node.function()->targetDefinition() );
-
-    const auto location = m_locationRegistry.get( functionDefintion->uid(), argumentValues );
-    const Update update{ updateValue, &node, m_agentId };
+    m_evaluateUpdateLocation = true;
```
Comment (emmanuel099) [inline src/execute/NumericExecutionPass.cpp]: Oh, this should be guarded (reset when function evaluation throws)

**Comment by ppaulweber** [comment]
So, if I correctly understand this PR, your are making the detection if a function on the FE level is the `program` function through a new method facility. Shall this be the same fashion then for `result` and `self`?

**Comment by emmanuel099** [comment]
Not required for `result` and `self`. Only the "is program" check is required during execution (and because of that cached in the `m_isProgram` boolean).

**Comment by ppaulweber** [comment]
:+1:

**Pull request merged**